app/models/backup.rb

```diff
@@ -50,8 +50,23 @@ def s3
 
   def upload_to_s3
     return unless s3
-    File.open(@path) do |file|
-      s3.upload(file, @filename)
+
+    if SiteSetting.s3_enable_encryption
+      if (s3_gnupg_public_key = SiteSetting.s3_gnupg_public_key).blank?
+        raise Discourse::SiteSettingMissing.new("s3_gnupg_public_key")
+      end
+
+      encrypted_filename = "#{@filename}.gpg"
+      encrypted_path = "/tmp/#{encrypted_filename}"
+
+      `echo '#{s3_gnupg_public_key}' | gpg --import`
+      gnupg_user_id = `echo '#{s3_gnupg_public_key}' | gpg --list-keys --with-colons | awk -F: '/^pub:/ { print $5 }'`.chomp
```

Comment:
I think you're doomed to hack around gpg's UI limitations... it is a fairly brutal tool.  The way you're doing things here, though, is likely to end in tears sooner or later, just because someone will change the GPG key used for encrypting their backups and your output will then not represent a single key...

You'll need to use a tempdir for storing the GPG config (or at least the public keyring), and then use `--no-default-keyring` and `--keyring /some/tmp/file` to ensure you *only* pick up the keyring file that contains the key you want to use for this session.